Elasticsearch Store Improvements

joemcelroy opened the pull request:
Todo:
- [x] Connection options (cloud, localhost url, es_connection) support
- [x] Logging support
- [x] Customisable field support
- [x] Distance Similarity support 
- [x] Metadata support
  - [x] Metadata Filter support 
- [x] Retrieval Strategies
  - [x] Approx
  - [x] Approx with Hybrid
  - [x] Exact
  - [x] Custom 
  - [x] ELSER (excluding hybrid as we are working on RRF support)
- [x] integration tests 
- [x] Documentation

👋 this is a contribution to improve Elasticsearch integration with Langchain. Its based loosely on the changes that are in master but with some notable changes:

## Package name & design improvements
The import name is now `ElasticsearchStore`, to aid discoverability of the VectorStore.

```py
## Before
from langchain.vectorstores.elastic_vector_search import ElasticVectorSearch, ElasticKnnSearch

## Now
from langchain.vectorstores.elasticsearch import ElasticsearchStore
```

## Retrieval Strategy support
Before we had a number of classes, depending on the strategy you wanted. `ElasticKnnSearch` for approx, `ElasticVectorSearch` for exact / brute force. 

With `ElasticsearchStore` we have retrieval strategies:

### Approx Example
Default strategy for the vast majority of developers who use Elasticsearch will be inferring the embeddings from outside of Elasticsearch. Uses KNN functionality of _search.

```py
        texts = ["foo", "bar", "baz"]
       docsearch = ElasticsearchStore.from_texts(
            texts,
            FakeEmbeddings(),
            es_url="http://localhost:9200",
            index_name="sample-index"
        )
        output = docsearch.similarity_search("foo", k=1)
```

### Approx, with hybrid
Developers who want to search, using both the embedding and the text bm25 match. Its simple to enable.

```py
 texts = ["foo", "bar", "baz"]
       docsearch = ElasticsearchStore.from_texts(
            texts,
            FakeEmbeddings(),
            es_url="http://localhost:9200",
            index_name="sample-index",
            strategy=ElasticsearchStore.ApproxRetrievalStrategy(hybrid=True)
        )
        output = docsearch.similarity_search("foo", k=1)
```

### Approx, with `query_model_id`
Developers who want to infer within Elasticsearch, using the model loaded in the ml node.

This relies on the developer to setup the pipeline and index if they wish to embed the text in Elasticsearch. Example of this in the test.

```py
 texts = ["foo", "bar", "baz"]
       docsearch = ElasticsearchStore.from_texts(
            texts,
            FakeEmbeddings(),
            es_url="http://localhost:9200",
            index_name="sample-index",
            strategy=ElasticsearchStore.ApproxRetrievalStrategy(
                query_model_id="sentence-transformers__all-minilm-l6-v2"
            ),
        )
        output = docsearch.similarity_search("foo", k=1)
```

### I want to provide my own custom Elasticsearch Query
You might want to have more control over the query, to perform multi-phase retrieval such as LTR, linearly boosting on document parameters like recently updated or geo-distance. You can do this with `custom_query_fn`

```py
        def my_custom_query(query_body: dict, query: str) -> dict:
            return {"query": {"match": {"text": {"query": "bar"}}}}

        texts = ["foo", "bar", "baz"]
        docsearch = ElasticsearchStore.from_texts(
            texts, FakeEmbeddings(), **elasticsearch_connection, index_name=index_name
        )
        docsearch.similarity_search("foo", k=1, custom_query=my_custom_query)

```

### Exact Example
Developers who have a small dataset in Elasticsearch, dont want the cost of indexing the dims vs tradeoff on cost at query time. Uses script_score.

```py
        texts = ["foo", "bar", "baz"]
       docsearch = ElasticsearchStore.from_texts(
            texts,
            FakeEmbeddings(),
            es_url="http://localhost:9200",
            index_name="sample-index",
            strategy=ElasticsearchStore.ExactRetrievalStrategy(),
        )
        output = docsearch.similarity_search("foo", k=1)
```

### ELSER Example
Elastic provides its own sparse vector model called ELSER. With these changes, its really easy to use. The vector store creates a pipeline and index thats setup for ELSER. All the developer needs to do is configure, ingest and query via langchain tooling.

```py
texts = ["foo", "bar", "baz"]
       docsearch = ElasticsearchStore.from_texts(
            texts,
            FakeEmbeddings(),
            es_url="http://localhost:9200",
            index_name="sample-index",
            strategy=ElasticsearchStore.SparseVectorStrategy(),
        )
        output = docsearch.similarity_search("foo", k=1)

```

## Architecture
In future, we can introduce new strategies and allow us to not break bwc as we evolve the index / query strategy. 

## Credit
On release, could you credit @elastic and @phoey1 please? Thank you!

joemcelroy committed:
introduce a new ElasticsearchStore

joemcelroy committed:
Merge remote-tracking branch 'upstream/master' into elasticsearch-improvements

baskaryan commented:
@joemcelroy feel free to ping when it's ready for review

joemcelroy committed:
update elasticsearch & tests

joemcelroy committed:
fix lint / test errors for python 3.8

joemcelroy commented:
Hey @baskaryan - its ready for review ❤️ 

@jeffvestal @baskaryan What are your thoughts with the existing Elasticsearch vectorstore class? I suggest we leave it in for now and will follow up with another PR to deprecate / remove based on feedback from these changes. The new changes are at feature parity with the older implementation (+ new features and test coverage) so should be relatively easy for developers to move over.

These changes also include fixing the integration tests for legacy vectorstore (the diff for vcr snapshots are to update this)

Im working on adding documentation now.

joemcelroy committed:
Merge remote-tracking branch 'upstream/master' into elasticsearch-improvements

joemcelroy committed:
added delete support + more test coverage for queries

joemcelroy committed:
documentation

joemcelroy committed:
omit elser test

joemcelroy committed:
replace with union to fix lint / test issue

baskaryan reviewed (commented):
overall looks pretty good to me, would be great to have someone from elastic review the elastic-specific logic!

i do think having a ElasticVectorSearch and ElasticsearchStore will be confusing. my suggestion would be to call this ElasticVectorSearchV2, move it to something like vectorstores/elastic_vector_search/base_v2.py (move the original to base.py), add a warning to ElasticVectorSearch on init that it's been deprecated and will soon be removed ,and update docs to use ElasticVectorSearchV2. And at some point in the future we can renamed ElasticVectorSearchV2 -> ElasticVectorSearch and drop the original.

thoughts @eyurtsev @jeffvestal?

jeffvestal commented:
@joemcelroy at one point we were also discussing renaming the file to `elasticsearch_vector_search` or whatever, but the key was moving `elastic` -> `elasticsearch` to clarify 

We should leave the existing elasticVectorSearch code around and add the warning. I would hate for people to update and have things break. But having a deprecation warning and some target date where it is removed would be ideal.

joemcelroy commented:
@baskaryan How do you typically manage breaking changes in langchain?

Having discussed internally at elastic, we want these package / class names to be Elasticsearch so it aids discoverability. Elastic isn't the right name for the vector store. As @jeffvestal mentioned, we did plan to move over to using Elasticsearch name. I think adding this duplication and deprecation strategy a good opportunity to do this, giving time for developers to move over. 

> We should leave the existing elasticVectorSearch code around and add the warning. I would hate for people to update and have things break. But having a deprecation warning and some target date where it is removed would be ideal.

Fully agree - we should leave the existing classes alone and allow developers to move across in their own time. Duplication is fine here vs breaking stuff. Its easy to move over and ElasticsearchStore has feature parity with both classes (vector and knn).

I am in the process of updating the notebook to this new way, removing the ElasticVectorSearch approach.

joemcelroy commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,863 @@
+        filter: Union[List[dict], None],
+        similarity: Union[DistanceStrategy, None],
+    ) -> Dict:
+        logger.error(f"similarity {similarity}")
```
Good spot! Fixed.

joemcelroy committed:
add documentation + fix error with wrong logging + add typing

joemcelroy committed:
remove unneeded pass

joemcelroy commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,863 @@
+        filter: List[dict],
+        similarity: Union[DistanceStrategy, None],
+    ) -> Dict:
+        pass
```
Yep thanks! Added docstring describing method

joemcelroy committed:
embeddings property

joemcelroy committed:
rollback elasticsearch updates

joemcelroy committed:
added integration notebook for new changes

joemcelroy committed:
Merge remote-tracking branch 'upstream/master' into elasticsearch-improvements

joemcelroy committed:
update title

joemcelroy committed:
rollback deps

joemcelroy committed:
upgrade instructions + deprecation notice

joemcelroy commented:
I've now updated the notebook which is used for the integration page. 

I've also added deprecated decorators for the old classes and upgrade instructions in the notebook.

Hope this is good enough @baskaryan :)

 

What

joemcelroy committed:
use warnings than deprecation decorator

joemcelroy commented:
ok shouldn't of used the decorators! Looking at other deprecations, they use the warnings. Updated it for that now.

Also let me know the error for vercel preview - i updated the notebook so likely caused by me.

joemcelroy committed:
fix lint + additional test coverage

joemcelroy committed:
Merge remote-tracking branch 'upstream/master' into elasticsearch-improvements

joemcelroy committed:
added some more examples to notebook

joemcelroy committed:
fixed format

eyurtsev reviewed (commented):
Code looks pretty good

To note: filter format follows native elasticsearch format -- more powerful, but I think different convention from other vectostores (not sure how standardized)

It's not standardized as far as I can tell between the vectorstores, though some of them use format that looks like this: https://python.langchain.com/docs/integrations/vectorstores/chroma#filtering-on-metadata

No suggestion at this time, just something to be aware of


Minor changes:
* Doc-strings to fit within the 88 character length
* Usage of `*` before named arguments to force caller code to use named arguments with defaults

joemcelroy commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,981 @@
+        Args:
+            texts: Iterable of strings to add to the vectorstore.
+            metadatas: Optional list of metadatas associated with the texts.
+            ids: Optional list of ids to associate with the texts.
```
thanks for spot on refresh_indices. kwargs isn't used, mainly here so matches the signature of the super class method. Is it ok if I leave it undocumented?

joemcelroy commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,981 @@
+        return ExactRetrievalStrategy()
+
+    @staticmethod
+    def ApproxRetrievalStrategy(
```
Thats it basically. Do you suggest another way for this?

eyurtsev replied:
Nope this is great -- wanted to make sure I understood the reason

joemcelroy commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,863 @@
+        filter: List[dict],
+        similarity: Union[DistanceStrategy, None],
+    ) -> Dict:
+        pass
```
thanks! done in previous commit

joemcelroy commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,863 @@
+    ) -> Dict:
+        pass
+
+    def beforeIndexSetup(
```
thanks for spot!

joemcelroy committed:
code review fixes

joemcelroy committed:
format fixes

joemcelroy commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,985 @@
+    def __init__(
+        self,
+        query_model_id: Optional[str] = None,
+        hybrid: Optional[bool] = False,
```
I updated the elasticsearch notebook documenting all the options for every strategy + I added details on the static method for the strategy on the `hybrid` argument. I hope thats good enough!

joemcelroy commented on libs/langchain/tests/integration_tests/vectorstores/docker-compose/elasticsearch.yml:
```diff
@@ -2,25 +2,33 @@ version: "3"
       - xpack.security.enabled=false # security has been disabled, so no login or password is required.
       - xpack.security.http.ssl.enabled=false
     ports:
       - "9200:9200"
```
Agreed. Will leave it for now but will change in future revision.

joemcelroy committed:
update documentation

joemcelroy committed:
updated the stringdoc to 88 chars limit

joemcelroy commented:
thank you @eyurtsev for your review. I committed the suggestions you made and a couple of follow up questions.

joemcelroy committed:
formatting

joemcelroy commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,863 @@
+        self,
+        query_vector: Union[List[float], None],
+        query: Union[str, None],
+        k: int,
```
i've updated for query but haven't adjusted anywhere else. For the strategies, i think its ok tho will see how the strategies evolve over time.

joemcelroy commented:
> To note: filter format follows native elasticsearch format -- more powerful, but I think different convention from other vectostores (not sure how standardized)
> 
> It's not standardized as far as I can tell between the vectorstores, though some of them use format that looks like this: https://python.langchain.com/docs/integrations/vectorstores/chroma#filtering-on-metadata

Yeh i arrived at the same thoughts - there doesn't seem to be a standardised way for filtering and Elasticsearch supports many more options over vector DBs so felt keep simple and use the native support.

joemcelroy committed:
fix commented out test assert

joemcelroy commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,981 @@
+        es_password: Optional[str] = None,
+        vector_query_field: str = "vector",
+        query_field: str = "text",
+        distance_strategy: Optional[DistanceStrategy] = DistanceStrategy.COSINE,
```
good suggestion. thanks!

joemcelroy commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,981 @@
+        if self.query_field not in fields:
+            fields.append(self.query_field)
+
+        fields.append("metadata")
```
yeh its a good point. For now if fields is None, then add metadata by default.

joemcelroy committed:
updated based on PR comments

joemcelroy commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,981 @@
+
+        if len(requests) > 0:
+            try:
+                bulk(self.client, requests)
```
done!

joemcelroy commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,981 @@
+        else:
+            # the search_type doesn't require inference, so we don't need to
+            # embed the texts.
+            self._create_index_if_not_exists(index_name=self.index_name)
```
done!

joemcelroy committed:
reformat

joemcelroy committed:
Merge remote-tracking branch 'upstream/master' into elasticsearch-improvements

jimczi commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,1054 @@
+
+        if len(requests) > 0:
+            try:
+                success, failed = bulk(self.client, requests, stats_only=True)
```
You can add `refresh=refresh_indices` in the parameter to avoid calling refresh explicitly below. It should be more performant since only the shards that receive updates will apply the refresh.

joemcelroy replied:
Yes nice little performance improvement! Done

jimczi commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,1054 @@
+                DistanceStrategy.EUCLIDEAN_DISTANCE,
+            ]
+        ] = None,
+        strategy: BaseRetrievalStrategy = ApproxRetrievalStrategy(),
```
Should it expose `number_of_shards` and `number_of_replicas` as parameters too ?

joemcelroy replied:
yeh :) Il follow this up in another PR. Probably a mappings / settings parameter thats used to extend.

jimczi commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,1054 @@
+        if self.embedding is not None:
+            # If no search_type requires inference, we use the provided
+            # embedding function to embed the texts.
+            embeddings = self.embedding.embed_documents(list(texts))
```
It's weird that texts is an iterable and the rest (ids and metadatas) are passed as lists. Not related to this PR though but I guess that add_texts is never called with large batch.

joemcelroy committed:
format

joemcelroy commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,1054 @@
+            raise ValueError(f"Similarity {similarity} not supported.")
+
+        return {
+            "mappings": {
```
Question: by default it should be excluded as the _source attribute is only `text_field` or explicitly specified fields via `fields` arg. Is making `vector_query_field` explicitly excluded with source exclude an improvement to this?

joemcelroy commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,1054 @@
+                "rank": {"rrf": {}},
+            }
+        else:
+            return {"knn": knn}
```
topic addressed in another comment. resolving convo

joemcelroy commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,1054 @@
+
+        # If hybrid, add a query to the knn query
+        # RRF is used to even the score from the knn query and text query
+        if self.hybrid:
```
My thoughts that case should never happen. I will check if langchain fails higher up the chain.

joemcelroy committed:
fix lint errors

joemcelroy committed:
lint and format fixes

joemcelroy commented:
i will get the hang of the lint / format soon 😂

joemcelroy committed:
ignore length check

baskaryan commented:
> i will get the hang of the lint / format soon 😂

try `make format` +  `make lint` (from libs/langchain) to fix+check things locally!

joemcelroy committed:
fix format

joemcelroy committed:
Merge remote-tracking branch 'upstream/master' into elasticsearch-improvements

joemcelroy commented:
thanks @baskaryan - hope this passes CI

baskaryan commented on libs/langchain/langchain/vectorstores/elasticsearch.py:
```diff
@@ -0,0 +1,1078 @@
+
+        if len(requests) > 0:
+            try:
+                success, failed = bulk(
```
hm i wonder what the behavior here should be. does failed indicate which texts failed or just the number that failed? should there be an option to raise an error if any inserts failed, and if so is there a way to undo the successful inserts?

can be addressed in subsequent pr's, but i do think the current implementation could lead to some unintuitive / hard to debug behavior

joemcelroy replied:
Thanks - do we have examples where other vectorstores do this on better messaging? I agree we could be better here. 

Elasticsearch doesn't support  any sort of rollback if one or more failed and wary of doing this within the client. If you agree, will create a ticket for improving the error messaging and logging the failed ids here.

baskaryan committed:
cr
